### PR TITLE
Apply event sourcing to multi instance

### LINKED ISSUE #6198: Apply event-sourcing to the BPMN multi-instance processor
**Description**

* apply the event-sourcing concept to the BPMN multi-instance processor

Reference:
* [ZEP: Multi-Instance](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md#multi-instance-activity)

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR applies event sourcing principles to the multi instance, by moving out all state changes to appliers (or temporarily to non-migrated elements' ELEMENT_ACTIVATING). The only difference in output is the new ACTIVATE, COMPLETE, and TERMINATE commands when dealing with multi instance bodies - other outputs are the same as before.

I had to fix a few issues, one related to the dirty hack in the reprocessing state machines (to correctly identify if a source of a sequence flow is migrated), and one related to finding the right event supplier (as the inner activity is always picked up instead of the multi instance on generic lookups).

NOTE: might be easier to look at the commits.

## Related issues

closes #6198 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by npepinpe** [comment]
I've added example tests, and I'll time box some time to see if I can easily extend the random generator - if not we can always fall back on the example tests.

**Comment by npepinpe** [comment]
I think I have an idea of how to extend the random workflow generator, but it would require some refactoring of the block builders (i.e. need to split away the activity builders, but service task sometimes builds only that, or sometimes builds a service task followed by a gateway - how do you split them?). We can probably leave it for later - or we can implement it for now but just, say, on receive task.

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReplayStateTest.java [hunk 1]
```diff
@@ -258,6 +259,156 @@
...
+                      .forEach(
+                          key -> engine.job().withKey(key).withVariable("result", 0).complete());
+
+                  return RecordingExporter.processInstanceRecords(
+                          ProcessInstanceIntent.ELEMENT_COMPLETED)
+                      .withProcessInstanceKey(piKey)
+                      .withElementType(BpmnElementType.PROCESS)
+                      .getFirst();
+                }),
+        testCase("sequential multi-instance service task")
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReplayStateTest.java:306]: This test lies :cake: 

In the process, the multi-instance should be sequential, not parallel.

**Review by saig0 — APPROVED** [review]
Great work :tada: 
Thanks to the small commits and the commit messages, it was easy to follow :+1: 

I've just one comment to a test case. Please have a look before merging.

File: engine/src/main/java/io/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java [hunk 2]
```diff
@@ -67,53 +67,22 @@ public MultiInstanceBodyProcessor(final BpmnBehaviors bpmnBehaviors) {
...
       final ExecutableMultiInstanceBody element, final BpmnElementContext context) {
-
     // verify that the input collection variable is present and valid
     readInputCollectionVariable(element, context)
-        .flatMap(ok -> eventSubscriptionBehavior.subscribeToEvents(element, context))
+        .flatMap(
+            inputCollection ->
+                eventSubscriptionBehavior
+                    .subscribeToEvents(element, context)
+                    .map(ok -> inputCollection))
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply engine/src/main/java/io/zeebe/engine/processing/bpmn/container/MultiInstanceBodyProcessor.java:78]: I almost put a comment that I know it looks weird and I'd hope for a better solution, but couldn't come up with one (without a specific `Try` variant of `Either` anyway) :see_no_evil:

File: engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReplayStateTest.java (hunk 1 above)
Comment (npepinpe) [reply engine/src/test/java/io/zeebe/engine/processing/streamprocessor/ReplayStateTest.java:306]: The dangers of copy pasting :see_no_evil:

**Pull request merged**